mutation_order_gibbs.py

```diff
@@ -360,3 +425,60 @@ def _get_denom_update(self, old_denominator, prev_feat_idxs, feat_mut_step):
             prev_exp_theta_sum = self.exp_theta_sum[prev_feat_idxs] if len(prev_feat_idxs) else self.exp_theta_num_cols
             new_denom = old_denominator - prev_exp_theta_sum - np.sum(old_feat_exp_theta_sums) + np.sum(new_feat_exp_theta_sums)
         return float(new_denom)
+
+    def _get_risk_update(self, base_risk_vec, prev_mut_pos, feat_mut_step, new_denom):
+        """
+        Similar to _get_denom_update but calculate the updated risk vector to use as residuals
+
+        @param base_risk_vec: the vector of summands from denominator from the previous mutation step weighted with the new denominator
+        @param prev_mut_pos: previously mutated position that won't contribute to the updated risk group
+        @param feat_mut_step: the features that differed for this next mutation step
+        @param new_denom: new denominator
+        """
+        for feat_pos, feat_idxs in feat_mut_step.neighbors_feat_new.iteritems():
+            # updated risk values for nearby positions
+            if self.feature_generator.num_feat_gens > 1:
+                if not self.per_target_model:
+                    base_risk_vec[feat_pos] = np.exp(self.theta[feat_idxs].sum()) / new_denom
+                else:
+                    base_risk_vec[feat_pos] = np.exp(self.theta[feat_idxs,0].sum() + self.theta[feat_idxs, 1:].sum(axis=0)) / new_denom
+            elif feat_idxs.size:
+                base_risk_vec[feat_pos] = self.exp_theta_sum[feat_idxs]/ new_denom
+        if len(prev_mut_pos):
+            # prev_mut_pos are now zero risk having mutated
+            base_risk_vec[prev_mut_pos] = 0.
+        return base_risk_vec
+
+    def _calculate_unoffset_residuals(self, acc_risks):
+        """
+        Get position-wise residuals
+
+        These are martingale residuals padded with nans where we ignored positions due to
+        mutations in the flanks.
+
+        @param acc_risks: numpy array of accumulated risk values
+
+        @return: indicator of whether a position mutated minus its accumulated risk; NaNs added to flanks and ambiguous positions
+        """
+
+        def _pad_vec_with_nan(vec):
+            """
+            Pad vector with NaN values both for the flanking positions and for the interior "n" values that were collapsed during processing
+            """
+            padded_vec = np.concatenate((
+                [np.nan] * self.obs_seq_mutation.left_position_offset,
```

Comment:
not a fan of the nans in the code. I'm worried that there could actually be a real nan vs. our own nan. Can you think of a way to distinguish between these two? Or should we just assume we'll never generate a real nan?